src/Common/src/System/Security/Cryptography/Asn1V2.cs

```diff
@@ -0,0 +1,5071 @@
+                {
+                    localLen += EndOfContentsEncodedLength;
+                }
+
+                totalLen += localLen;
+                cur = cur.Slice(localLen);
+            }
+
+            throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+        }
+
+        public ReadOnlyMemory<byte> PeekEncodedValue()
+        {
+            (Asn1Tag tag, int? length) = ReadTagAndLength(out int bytesRead);
+
+            if (length == null)
+            {
+                var tagLengthAndContents = SeekEndOfContents(_data, _ruleSet, bytesRead);
+                return Slice(_data, 0, tagLengthAndContents.Length + EndOfContentsEncodedLength);
+            }
+
+            return Slice(_data, 0, bytesRead + length.Value);
+        }
+
+        public ReadOnlyMemory<byte> PeekContentBytes()
+        {
+            (Asn1Tag tag, int? length) = ReadTagAndLength(out int bytesRead);
+
+            if (length == null)
+            {
+                return SeekEndOfContents(_data.Slice(bytesRead), _ruleSet);
+            }
+
+            return Slice(_data, bytesRead, length.Value);
+        }
+
+        public void SkipValue()
+        {
+            GetEncodedValue();
+        }
+
+        public ReadOnlyMemory<byte> GetEncodedValue()
+        {
+            ReadOnlyMemory<byte> encodedValue = PeekEncodedValue();
+            _data = _data.Slice(encodedValue.Length);
+            return encodedValue;
+        }
+
+        private static bool ReadBooleanValue(
+            ReadOnlySpan<byte> source,
+            AsnEncodingRules ruleSet)
+        {
+            // T-REC-X.690-201508 sec 8.2.1
+            if (source.Length != 1)
+            {
+                throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+            }
+
+            byte val = source[0];
+
+            // T-REC-X.690-201508 sec 8.2.2
+            if (val == 0)
+            {
+                return false;
+            }
+
+            // T-REC-X.690-201508 sec 11.1
+            if (val != 0xFF && (ruleSet == AsnEncodingRules.DER || ruleSet == AsnEncodingRules.CER))
+            {
+                throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+            }
+
+            return true;
+        }
+
+        public bool ReadBoolean() => ReadBoolean(Asn1Tag.Boolean);
+
+        public bool ReadBoolean(Asn1Tag expectedTag)
+        {
+            (Asn1Tag tag, int? length) = ReadTagAndLength(out int headerLength);
+            CheckExpectedTag(tag, expectedTag, UniversalTagNumber.Boolean);
+
+            // T-REC-X.690-201508 sec 8.2.1
+            if (tag.IsConstructed)
+            {
+                throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+            }
+
+            bool value = ReadBooleanValue(
+                Slice(_data, headerLength, length.Value).Span,
+                _ruleSet);
+
+            _data = _data.Slice(headerLength + length.Value);
```

Comment:
We should evaluate how operations like this affect the lifetime of _data (especially if this ever becomes a public API where any method could be the last usage of _data). It's possible this could be reordered such that _data gets collected while the Span is still in use.